posts/2017-02-13-more-types-than-classes.md

```diff
@@ -0,0 +1,702 @@
+
+The colloquial “runtime type”, a fair approximation of “class”, makes
+it tempting to equate _types_ with “classes, interfaces, traits, that
+sort of thing”, which I will name _classes_ for the rest of this
+article. But they aren’t the same. The type system is much richer and
+more interesting than the class system, even in Java.
+
+To appreciate this richness, we must stop thinking of types as classes
+and stop drawing conclusions from that weak analogy. Luckily, the
+compiler will readily reveal how unlike classes types are, if we ask
+it some simple questions.
+
+## One value with class, many variables with type
+
+```scala
+val greeting: String = "hi there!"
+```
+
+Here I have constructed a `String` and assigned it to a variable. (I
+have also constructed the `char` array in the `String` and various
+other details, but immediately handed those off to the `String` and
+forgotten about them.) This value has class `String`. It has several
+classes, really.
+
+1. `String`
+2. `java.io.Serializable`
+3. `CharSequence`
+4. `Comparable[String]`
+5. `Object`/`AnyRef`
+
+That seems like a lot of classes for one value. And they are genuine
+classes of `greeting`, though 2-5 are all implied by #1.
+
+`greeting` also has all five of these _types_. We can ask the compiler
+to verify that this _type_ truth holds, entirely separately from the
+class truth.
+
+```scala
+scala> (greeting: String, greeting: java.io.Serializable,
+        greeting: CharSequence, greeting: Comparable[String],
+        greeting: AnyRef)
+res3: (String, java.io.Serializable, CharSequence,
+       Comparable[String], AnyRef) =
+  (hi there!,hi there!,hi there!,hi there!,hi there!)
+```
+
+So we have exhausted the classes, but aren’t quite done with types.
+
+```scala
+scala> greeting: greeting.type
+res4: greeting.type = hi there!
+```
+
+`greeting.type` is not like the other five types we just tested. It is
+a strict subtype of `String`, and has no class with the same name.
+
+```scala
+// If and only if call compiles, A is a subtype of B.
+def conformance[A, B >: A]: Unit = ()
+
+scala> conformance[greeting.type, String]
+
+scala> conformance[String, greeting.type]
+<console>:14: error: type arguments [String,greeting.type] do not conform
+              to method conformance's type parameter bounds [A,B >: A]
+```
+
+Fine, we can accept that object identity is represented at the type
+level without our universe imploding, by inventing the theory that
+this is about object identity; hold on, though:
+
+```scala
+scala> val salutation = greeting
+salutation: String = hi there!
+```
+
+Fine, `salutation` is just another name for `greeting`, right?
+
+```scala
+scala> conformance[salutation.type, String]
+
+scala> implicitly[greeting.type =:= salutation.type]
+<console>:14: error: Cannot prove that greeting.type =:= salutation.type.
+```
+
+Now we have seven. I’ll spare you spelling out the induction: each new
+variable defined like `salutation` will yield a new alias with a
```

Comment:
You might want to clarify that you are using *variable* here in the sense of binding (`val`), not in the sense of mutation (`var`). Maybe this is overly pedantic since the examples all use `val` but I could imagine someone getting the wrong idea.